### PR TITLE
fix: update and refactor to aws sdk v3

### PR DESCRIPTION
**Pull request by ms-fcabrera**
Update the aws sdk to v3, refactor due to DynamoDb new implementations

Evidence:
image

### CODE REVIEW

File: src/app/UserService.ts [hunk 1]
```diff
@@ -1,77 +1,69 @@
...
 
-    const item = result.Item || key;
-    return this.mapUser(item);
+    const result = await this._dbClient.get({
+      TableName: this._userTableName,
+      Key: key,
+    });
+
+    const item = result.Item || { email };
+    return this.mapUser(item as Record<string, unknown>);
```
Comment (ms-fcabrera) [inline src/app/UserService.ts:28]: I'm not sure this is the last thing we got to in the pair programming session.

Comment (andresmoschini) [reply src/app/UserService.ts:28]: Mmm, as I understand, the cast should not be necessary.

File: src/app/User.test.ts [hunk 2]
```diff
@@ -13,8 +13,10 @@ describe(UserService.name, () => {
 
       // Assert
       expect(dbClientDouble.get).toHaveBeenCalledWith({
-        Key: { email: "email1" },
-        TableName,
+        TableName: TableName,
+        Key: {
+          email: "email1",
+        },
```
Comment (andresmoschini) [inline src/app/User.test.ts:19]: This change is not necessary.

File: src/app/User.test.ts [hunk 3]
```diff
@@ -58,20 +60,21 @@ describe(UserService.name, () => {
...
       await sut.registerReportsSectionVisit("email1", date);
 
       // Assert
       expect(dbClientDouble.update).toHaveBeenCalledWith({
-        TableName,
-        Key: { email: "email1" },
+        TableName: TableName,
+        Item: {
+          email: "email1",
+        },
```
Comment (andresmoschini) [inline src/app/User.test.ts:73]: This change is not necessary

File: src/app/User.test.ts [hunk 4]
```diff
@@ -58,20 +60,21 @@ describe(UserService.name, () => {
     it("Should pass the right data to db", async () => {
       // Arrange
       const date = new Date("Date Mon Oct 24 2022 17:21:08 GMT-0300");
-      const expectedValue = "2022-10-24T20:21:08.000Z";
```
Comment (andresmoschini) [inline src/app/User.test.ts:61]: Removing redundant code is good, but it would be nice do it in another commit.

Comment (andresmoschini) [reply src/app/User.test.ts:61]: I think that this code should not be removed. In any case, it should be part of another commit because it is not part of "update and refactor to aws sdk v3"

File: src/app/User.test.ts [hunk 5]
```diff
@@ -58,20 +60,21 @@ describe(UserService.name, () => {
...
-        Key: { email: "email1" },
+        TableName: TableName,
+        Item: {
+          email: "email1",
+        },
         UpdateExpression:
           "set reportsSectionLastVisit = :reportsSectionLastVisit",
         ExpressionAttributeValues: {
-          ":reportsSectionLastVisit": expectedValue,
+          ":reportsSectionLastVisit": date.toISOString(),
```
Comment (andresmoschini) [inline src/app/User.test.ts:77]: Oh, I prefer the original implementation, because it was explicit about what was the value format (exactly `2022-10-24T20:21:08.000Z`), without the need to go to [toISOString documentation](https://developer.mozilla.org/es/docs/Web/JavaScript/Reference/Global_Objects/Date/toISOString)

File: src/app/UserService.ts [hunk 6]
```diff
@@ -1,77 +1,76 @@
...
-        ExpressionAttributeValues: {
-          ":reportsSectionLastVisit": date.toISOString(),
-        },
-      })
-      .promise();
+    await this._dbClient.update({
+      TableName: this._userTableName,
+      Item: {
+        email,
+      },
```
Comment (andresmoschini) [inline src/app/UserService.ts:39]: In my opinion, the previous name `Key` fitted better with the meaning.

File: src/doppler-security/JwtFilter.test.ts [hunk 7]
```diff
@@ -196,7 +195,7 @@ describe(JwtFilter.name, () => {
       it(`Apply should fail for ${scenario}`, async () => {
         const sut = new JwtFilter({ jwtVerifier });
         const result = await sut.apply(
-          { headers } as unknown as APIGatewayProxyEvent,
+          { headers } as unknown,
```
Comment (andresmoschini) [inline src/doppler-security/JwtFilter.test.ts:198]: El cast a unknown creo que es redundante.

File: src/app/User.test.ts (hunk 4 above)
Comment (ms-fcabrera) [reply src/app/User.test.ts:61]: 👍

File: src/app/User.test.ts (hunk 5 above)
Comment (ms-fcabrera) [reply src/app/User.test.ts:77]: 👍

File: src/app/UserService.ts (hunk 6 above)
Comment (ms-fcabrera) [reply src/app/UserService.ts:39]: I was having a typo with the selected commands, I was using `PutCommand` that is for creating new items in dynamo, and that comand needs a parameter `Item` (we might as well have mapped it). The command needed for updating an existing item was `UpdateCommand` that instead of `Item` receives `Key` as a parameter, so actually all methods will be using `Key`

File: src/app/User.test.ts (hunk 3 above)
Comment (ms-fcabrera) [reply src/app/User.test.ts:73]: 👍

File: src/app/User.test.ts (hunk 2 above)
Comment (ms-fcabrera) [reply src/app/User.test.ts:19]: 👍

File: src/app/DummyDbClient.ts [hunk 8]
```diff
@@ -1,19 +1,20 @@
...
+    get: (key: Record<string, any>) =>
       createPromiseWrapperWithDelay(delay, () => ({
         Item: {
           email: key.email,
         },
       })),
+    delete: (key: Record<string, any>) => createPromiseWrapperWithDelay(delay),
   };
-  return dbClientDouble as unknown as DynamoDB.DocumentClient;
+  return dbClientDouble as unknown as DynamoDocumentClient;
```
Comment (andresmoschini) [inline src/app/DummyDbClient.ts:19]: This cast should not be necessary anymore because now you are implementing the complete `DynamoDocumentClient` interface. If you need it, probably means that this implementation is wrong or has missing parts.

File: src/doppler-security/JwtFilter.test.ts (hunk 7 above)
Comment (ms-fcabrera) [reply src/doppler-security/JwtFilter.test.ts]: 👍

File: src/app/UserService.ts (hunk 1 above)
Comment (ms-fcabrera) [reply src/app/UserService.ts:28]: 👍
